Recipes/Telemetry/WebApiTelemetryExtensions.cs

```diff
@@ -33,23 +36,52 @@ internal static class WebApiTelemetryExtensions
                 telemetry.Succeeded = response.IsSuccessStatusCode;
                 telemetry.HttpStatusCode = response.StatusCode;
 
-                if (response.RequestMessage != null)
+                var request = response.RequestMessage;
+                if (request != null)
                 {
-                    telemetry.RequestUri = response.RequestMessage.RequestUri;
-
-                    var callerIpAddress = response.RequestMessage.CallerIpAddress();
-
-                    if (!string.IsNullOrWhiteSpace(callerIpAddress))
-                    {
-                        telemetry.IsIncomingRequest(true);
-                        telemetry.CallerIpAddress(callerIpAddress);
-                    }
+                    telemetry.RequestUri = request.RequestUri;
+                    telemetry.WithCallerIpAddressBasedOn(request);
+                    telemetry.WithOperationNameBasedOn(request);
                 }
             }
 
             return telemetry;
         }
 
+        private static void WithCallerIpAddressBasedOn(this Telemetry telemetry, HttpRequestMessage request)
+        {
+            var callerIpAddress = request.CallerIpAddress();
+
+            if (!string.IsNullOrWhiteSpace(callerIpAddress))
+            {
+                telemetry.IsIncomingRequest(true);
+                telemetry.CallerIpAddress(callerIpAddress);
+            }
+        }
+
+        private static void WithOperationNameBasedOn(this Telemetry telemetry, HttpRequestMessage request)
+        {
+            if (request.HasActionDescriptor())
+            {
+                telemetry.OperationName = request.GetActionDescriptor().ActionName;
+            }
+            else
+            {
+                var leftPart = request.RequestUri.GetLeftPart(UriPartial.Authority);
+                telemetry.OperationName = new Uri(leftPart).MakeRelativeUri(request.RequestUri).ToString();
+            }
+        }
+
+        private static bool HasActionDescriptor(this HttpRequestMessage request)
+        {
+            return request.GetActionDescriptor() != null;
+        }
+
+        public static Dictionary<string, string> ToDictionary(this HttpHeaders httpHeaders)
+        {

```

Comment:
Actually this extension method is used by TelemetryAndLoggingMessageHandler which is in our Recipes project. I did rename it based on a feedba